packages/dd-trace/src/lambda/handler.js

```diff
@@ -86,6 +86,10 @@ exports.datadog = function datadog (lambdaHandler) {
     const context = extractContext(args)
 
     checkTimeout(context)
-    return lambdaHandler.apply(this, args).then((res) => { clearTimeout(__lambdaTimeout); return res })
+    if (lambdaHandler.constructor.name === 'AsyncFunction') {
+      return Promise.resolve(lambdaHandler.apply(this, args))
+        .then((res) => { clearTimeout(__lambdaTimeout); return res })
+    }
+    return lambdaHandler.apply(this, args)
```

Comment:
Is it valid for a plain function to return a `new Promise(...)` or a thenable? Perhaps should be doing something more like:

```suggestion
    const result = lambdaHandler.apply(this, args)
    if (result && typeof result.then === 'function') {
      return result.then((res) => {
        clearTimeout(__lambdaTimeout)
        return res
      })
    }
    return result
```